sdk/java/dagger-java-annotation-processor/src/main/java/io/dagger/annotation/processor/DaggerType.java

```diff
@@ -0,0 +1,293 @@
+    }
+
+    switch (name) {
+      case "void" -> {
+        return new Kind("void", true);
+      }
+      case "boolean" -> {
+        return new Kind("boolean", false);
+      }
+      case "int", "long", "short", "byte" -> {
+        return new Kind("int", false);
+      }
+      case "float", "double" -> {
+        return new Kind("float", false);
+      }
+    }
+
+    if (name.startsWith("java.util.List<")) {
+      return new List(name.substring("java.util.List<".length(), name.length() - 1));
+    }
+    if (!kindName.isEmpty() && kindName.equals(TypeKind.ARRAY.name())) {
+      // in that case the type name is com.example.Type[]
+      // so we remove the [] to get the underlying type
+      return new Array(name.substring(0, name.length() - 2));
+    }
+
+    if (name.startsWith("java.util.Optional<")) {
+      return of(name.substring("java.util.Optional<".length(), name.length() - 1));
+    }
+
+    try {
+      var clazz = Class.forName(name);
+      if (clazz.isEnum()) {
+        return new Enum(name, name.substring(name.lastIndexOf('.') + 1));
+      } else if (io.dagger.client.Scalar.class.isAssignableFrom(clazz)) {
+        return new Scalar(name, name.substring(name.lastIndexOf('.') + 1));
+      }
+    } catch (ClassNotFoundException e) {
+      // we are ignoring here any ClassNotFoundException
+      // not ideal, but we only want to know if it's an enum or a Scalar
+    }
+
+    try {
+      if (name.startsWith("java.lang.")) {
+        String simpleName = name.substring(name.lastIndexOf('.') + 1);
+        // check if it exists as TypeDefKind
+        TypeDefKind.valueOf("%s_KIND".formatted(simpleName.toUpperCase()));
+        return new Kind(simpleName, false);
+      }
+    } catch (IllegalArgumentException e) {
+      // it means valueOf failed
+    }
+
+    return new Object(name, name.substring(name.lastIndexOf('.') + 1));
+  }
+
+  public static DaggerType of(String name) {
+    if (name.equals("boolean")) {
+      return of(new TypeInfo(name, TypeKind.BOOLEAN.name()));
+    } else if (name.equals("byte")) {
+      return of(new TypeInfo(name, TypeKind.BYTE.name()));
+    } else if (name.equals("short")) {
+      return of(new TypeInfo(name, TypeKind.SHORT.name()));
+    } else if (name.equals("int")) {
+      return of(new TypeInfo(name, TypeKind.INT.name()));
+    } else if (name.equals("long")) {
+      return of(new TypeInfo(name, TypeKind.LONG.name()));
+    } else if (name.equals("char")) {
+      return of(new TypeInfo(name, TypeKind.CHAR.name()));
+    } else if (name.equals("float")) {
+      return of(new TypeInfo(name, TypeKind.FLOAT.name()));
+    } else if (name.equals("double")) {
+      return of(new TypeInfo(name, TypeKind.DOUBLE.name()));
+    } else if (name.equals("void")) {
+      return of(new TypeInfo(name, TypeKind.VOID.name()));
+    } else if (name.endsWith("[]")) {
+      return of(new TypeInfo(name, TypeKind.ARRAY.name()));
+    } else {
+      return of(new TypeInfo(name, ""));
+    }
```

Comment:
Nit: Could be a switch case too with an if on default to check if it endsWith `[]`?

Reply:
Yeah, I thought about that several times but never decided which one to pick 😅 
I'll give a new try

Reply:
done 👍